### PR TITLE
scripts: runners: bossac: Add legacy mode and fix arduino_nano_33_ble

### LINKED ISSUE #33523: Bossac runner flashes at an incorrect offset
**Describe the bug**
When flashing the arduino-nano-33-ble-sense using west the bossac runner passes --offset even though bossac seem to account for the elf offset.
This results in a "double" offset of 128kB instead of the 64kB offset where the bootloader starts execution.

To workaround this issue I can flash the device using only the bossac tool, without passing the --offset flag.

The result of the west flash is that the first 64kB of the old image `[0x10000, 0x20000]` still remain.

**To Reproduce**
```
west build -p -b arduino_nano_33_ble zephyr/samples/hello_world
west flash --bossac=/home/estrom/.arduino15/packages/arduino/tools/bossac/1.9.1-arduino2/bossac
```

If your binaries are sub 64kB then the previous commands result in a "NOP",
if you dump flash you can see the old program at 0x10000 and the new program with a 0x20000 offset.

**Expected behavior**
Hello world UART output on the TX and RX pins.

**Impact**
Unable to flash board using west

**Logs and console output**
The bare bossac flash without --offset passed to it,
```
017C90  3C 01 40 01 44 01 48 01 4C 01 50 01 61 72 64 75  |<.@.D.H.L.P.ardu|
017CA0  69 6E 6F 5F 6E 61 6E 6F 5F 33 33 5F 62 6C 65 00  |ino_nano_33_ble.|
017CB0  48 65 6C 6C 6F 20 57 6F 72 6C 64 21 20 25 73 0A  |Hello World! %s.|
```

The west bossac flash, which is passed `-o 65536`,
```
027C90  3C 01 40 01 44 01 48 01 4C 01 50 01 61 72 64 75  |<.@.D.H.L.P.ardu|
027CA0  69 6E 6F 5F 6E 61 6E 6F 5F 33 33 5F 62 6C 65 00  |ino_nano_33_ble.|
027CB0  48 65 6C 6C 6F 20 42 61 6A 73 74 69 61 6E 21 20  |Hello Bajstian! |
```

**Environment (please complete the following information):**
Linux, Debian Testing
zephyr-sdk 0.12.3
bossac "1.9.1-arduino2" shipped with the arduino IDE
zephyr v2.5-branch

### PR DESCRIPTION
**Pull request by nandojve**
Add compatibility mode with old sam-ba flash bootloaders that don't have offset capabilities.  These bootloaders flash to a pre-defined flash region.  At end, bossac will suppress --offset parameter.

This fixes arduino_nano_33_ble board that uses an old bootloader version.  The bootloader always relocate code due to flash bootloader.  This imply a wrong behaviour when using --offset.

Fixes #33523

Signed-off-by: Gerson Fernando Budke <nandojve@gmail.com>

### CODE REVIEW

**Review by eHammarstrom — APPROVED** [review]

**Review by mbolivar-nordic — APPROVED** [review]

**Comment by nandojve** [comment]
buildkite fail is unrelated

```
HEAD is now at 52dc9c5ad Merge pull request #10 from nashif/update_may
--
  | fatal: update_ref failed for ref 'refs/heads/manifest-rev': cannot lock ref 'refs/heads/manifest-rev': Unable to create '/var/lib/buildkite-agent/zephyr-module-cache/modules/lib/tinycbor/.git/refs/heads/manifest-rev.lock': File exists.
  |  
  | Another git process seems to be running in this repository, e.g.
  | an editor opened by 'git commit'. Please make sure all processes
  | are terminated then try again. If it still fails, a git process
  | may have crashed in this repository earlier:
  | remove the file manually to continue.
  | HEAD is now at 3e9a49d cmake: Fix conditional in root CMakeLists.txt
  | HEAD is now at 9e4498d lfs_util: fix log declaration when logging is disabled
  | HEAD is now at 75e6715 zephyr: move platform.h to zephyr tree
  | HEAD is now at a47e326 HW_models: NRF_RTC: Fix compare event missed when interrupt disabled
  | HEAD is now at 2f04b61 zephyr: add SoC overlay for intel_tgl_adsp
  | HEAD is now at 7dd56fc Add FIXME note to handle incorrect LL/CON/MAS/BV-27-C implementation
  | HEAD is now at 51cdecd6 trusted-firmware-m: pass compiler path to STM32L552 postbuild script
  | HEAD is now at 4aa516e Bump versions for v1.7.2
  | HEAD is now at d148bd2 Add zephyr module support
  | HEAD is now at dc70a45a7cc tensorflow: add tensorflow 2.4.1 as a zephyr module
  | ERROR: update failed for project tinycbor
  | touch: cannot touch 'twister-out/twister.xml': No such file or directory
```

**Comment by WilliamGFish** [comment]
I have applied the PR and it appears to work. The USB port on the nano 33 still doesn't seem to function.
I have used both versions of 1.9.1 the shumatech.com and Arduino IDE version, they both flash successfully.
Without additional equipment I haven't checked if the erase function is working correctly.

I have changed some of the switches to provide better feedback in the bossac  west runner script: (scripts\west_commands\runners\bossac.py - ln 159)
```
        cmd_flash = [self.bossac, '-d', '-i', '-U', '-p', self.port, '-R', '-e', '-w', '-v',
                      self.cfg.bin_file]
```
and amended the Windows error check: (ln 194)
```
        if platform.system() == 'Windows' and self.port == DEFAULT_BOSSAC_PORT:
            msg = 'CAUTION: BOSSAC runner on Windows requires --bossac-port defined, ie COM7!'
            raise RuntimeError(msg)
```

To get west to flash to flash to boards in Windows I am using:
`west -v flash --bossac="bossac.exe" --bossac-port="COM7"`
 
You want to roll these amendments in your PR (and give me  a mention)

File: Kconfig.zephyr
Comment (WilliamGFish) [inline Kconfig.zephyr:556]: Maybe you should rename to BOOTLOADER_BOSSA_MBED ?

**Review by mbolivar-nordic — APPROVED** [review]
If the 'legacy' bootloader is specific to MBED, I agree with @WilliamGFish about the proposed rename. Note I don't know this hardware so I will leave it to you to decide @nandojve. Otherwise this LGTM.

**Comment by nandojve** [comment]
> You want to roll these amendments in your PR (and give me a mention)

@WilliamGFish , thank you for you help with this. However, this patch add a mode to be compatible with some legacy bootloader/bossac version which fixes, by consequence, a problem with a current board in the tree. Because of that, I prefer not mix with any other specific enhancement.

Feel free to open a PR with your suggestion. It'll be nice reviewed.

**Comment by WilliamGFish** [comment]
> > > > You want to roll these amendments in your PR (and give me a mention)
> > > 
> > > 
> > > @WilliamGFish , thank you for you help with this. However, this patch add a mode to be compatible with some legacy bootloader/bossac version which fixes, by consequence, a problem with a current board in the tree. Because of that, I prefer not mix with any other specific enhancement.
> > > Feel free to open a PR with your suggestion. It'll be nice reviewed.
> > 
> > 
> > I can't comment on the Legacy boards but the offset also effects the nano 33 (nRF based) boards. It seems odd that it would work with the Adafruit bootloader but not the MBed?
> > LGTM..!!
> 
> It is not about boards it is about bootloader/tool not following what was defined by shumatech/BOSSA for non-ROM bootloaders.
> 
> _I'm fine with requiring that it be explicitly specified for non-ROM bootloaders like the SAMD's._
> _Originally posted by @shumatech in [shumatech/BOSSA#79 (comment)](https://github.com/shumatech/BOSSA/issues/79#issuecomment-415996880)_
> 
> Legacy means a system that not follow BOSSA >= 1.9 specs.

Happy either way, as it works with both the ShumaTech and Arduino versions of BOSSAC. 

The bigger question is do we stick with the ShumaTech version are adopt the Arduino branch of BOSSAC, the later appears to be more actively maintained? 
Eg: They have already been addressing the zero offset problems in their version. https://github.com/arduino/BOSSA/commit/89f3556a761833522cd93c199581265ad689310b

**Comment by Stefan-Schmidt** [comment]
Is this blocked by the approval or the CI failure?

**Comment by WilliamGFish** [comment]
> Is this blocked by the approval or the CI failure?

It looks like both...

**Comment by nandojve** [comment]
rebase

**Review by carlescufi — APPROVED** [review]

**Pull request merged**
